### PR TITLE
Merge/alns daily rest

### PR DESCRIPTION
**Pull request by hkgrov**
Mainly introduces daily rest restriction to the ALNS. 
In addition it introduces some improvements to the week repair operators. 
Lastly it introduces a new local search operator that focuses on making illegal solution breaking contracted hours legal.

### CODE REVIEW

File: heuristic/alns.py [hunk 1]
```diff
@@ -158,35 +169,71 @@ def __init__(self, state, data, criterion):
             self.shifts_at_day,
         )
 
+       
+
+
+
+        # remove_worst_week = partial(worst_week_removal, self.competencies, self.time_periods_in_week, self.combined_time_periods_in_week, self.employees, self.weeks, self.L_C_D, self.shifts_in_week, self.t_covered_by_shift)
```
Comment (ekallevik) [inline heuristic/alns.py:176]: Assume that all of these should be reintroduced?

File: heuristic/alns.py [hunk 2]
```diff
@@ -158,35 +169,71 @@ def __init__(self, state, data, criterion):
...
             destroy_set, destroy_specific_set = destroy_operator(candidate_solution)
             repair_set = repair_operator(candidate_solution, destroy_set, destroy_specific_set)
 
             self.calculate_objective(candidate_solution, destroy_set, repair_set)
-            self.consider_candidate_and_update_weights(
-                candidate_solution, destroy_operator_id, repair_operator_id
-            )
+            self.consider_candidate_and_update_weights(candidate_solution, destroy_operator_id, repair_operator_id)
 
-            candidate_solution = self.current_solution.copy()
```
Comment (ekallevik) [inline heuristic/alns.py:187]: Is this on purpose?

File: heuristic/heuristic_calculations.py [hunk 3]
```diff
@@ -107,6 +88,7 @@ def calculate_isolated_working_days(data, x):
             )
     return isolated_working_days
 
+    return isolated_working_days
```
Comment (ekallevik) [inline heuristic/heuristic_calculations.py:91]: This line should be removed

File: heuristic/repair_operators.py [hunk 4]
```diff
@@ -38,21 +38,24 @@ def worst_week_repair(shifts_in_week, competencies, t_covered_by_shift, employee
...
-        )
+        deviation_from_demand = -sum(state.soft_vars["deviation_from_ideal_demand"][c,t] 
+                                    for c in shift[2]
+                                    for t in t_covered_by_shift[shift[0], shift[1]] 
+                                    if (c, t) in state.soft_vars["deviation_from_ideal_demand"] 
+                                    if state.soft_vars["deviation_from_ideal_demand"][c,t] < 0)
 
-        if deviation_from_demand <= 6:
-            return repair_set
+#        deviation_from_demand = -sum(state.soft_vars["deviation_from_ideal_demand"][c,t] for c in competencies for t in time_periods_in_week[c, week[0]] if state.soft_vars["deviation_from_ideal_demand"][c,t] < 0)
```
Comment (ekallevik) [inline heuristic/repair_operators.py:51]: Could this be removed?

File: heuristic/repair_operators.py [hunk 5]
```diff
@@ -38,21 +38,24 @@ def worst_week_repair(shifts_in_week, competencies, t_covered_by_shift, employee
...
 
+        if(deviation_from_demand <= 6):
+            print("Deviation from demand: " + str(deviation_from_demand))
```
Comment (ekallevik) [inline heuristic/repair_operators.py:54]: Use `logger.info()`

File: heuristic/repair_operators.py [hunk 6]
```diff
@@ -69,16 +72,19 @@ def worst_week_repair(shifts_in_week, competencies, t_covered_by_shift, employee
...
-            )
-            for (competency_level, e) in employee_with_competency_combination[competencies_needed]
-            if (sum(state.x[e, t, v] for t, v in shifts_at_day[int(shift[0] / 24)])) == 0
-        }
+        
+        deviation_contracted_hours = {e: (sum(state.soft_vars["deviation_contracted_hours"][e,j] for j in weeks)  
+                                    - (competency_level - len(competencies_needed)))
+                                    + (20 if (sum(state.soft_vars["deviation_contracted_hours"][e,j] for j in weeks) - shift[1] >= 8.5) else 0)
+                                    + (20 if (sum(state.soft_vars["deviation_contracted_hours"][e,j] for j in weeks) - shift[1] <= 17) else 0)
+                                    + (100 if (sum(state.soft_vars["deviation_contracted_hours"][e,j] for j in weeks) - shift[1]  == 0) else 0)
```
Comment (ekallevik) [inline heuristic/repair_operators.py:80]: What is the intention here? To reward: 
- +20 if the deviation is longer than 17
- +20 if the deviation is shorter than 8.5
- +40 if the deviation is between 8.5 and 17
- +100 if the deviation is 0
? 

Why exactly this reward scheme?

File: heuristic/repair_operators.py [hunk 7]
```diff
@@ -69,16 +72,19 @@ def worst_week_repair(shifts_in_week, competencies, t_covered_by_shift, employee
...
+                                    for (competency_level, e) in employee_with_competency_combination[competencies_needed] 
+                                    if (sum(state.x[e,t,v] for t,v in shifts_at_day[int(shift[0]/24)])) == 0}
+
+        #print("DEV: " + str(deviation_contracted_hours))
```
Comment (ekallevik) [inline heuristic/repair_operators.py:84]: Remove

File: heuristic/alns.py (hunk 2 above)
Comment (hkgrov) [reply heuristic/alns.py:187]: Yes it is on purpose. I believe through another merge this has been introduced by mistake. Why would you have another copy of the state at the bottom when it the next line in the loop says the exact same thing?

File: heuristic/repair_operators.py (hunk 6 above)
Comment (hkgrov) [reply heuristic/repair_operators.py:80]: It is mostly a test. Did improve the solution, but should be improved upon.

Comment (ekallevik) [reply heuristic/repair_operators.py:76]: I see! There is definitely potential for improvement here, both with regards to the actual weighting and performance (4 loops for calculating the sum seems 2 more than needed). If you feel confident enough about this we can merge it now, but then we need to come back to this later. Can you add a Trello-card for this?

File: heuristic/repair_operators.py [hunk 8]
```diff
@@ -269,19 +271,14 @@ def worst_employee_repair(competencies, t_covered_by_shift, employee_with_compet
...
-            for c in competencies_2
-            for t in t_covered_by_shift[shift[0], shift[1]]
-            if (c, t) in state.soft_vars["deviation_from_ideal_demand"]
-            if state.soft_vars["deviation_from_ideal_demand"][c, t] < 0
-        )
-        if deviation_from_demand < 6:
+                
+        deviation_from_demand = -sum(state.soft_vars["deviation_from_ideal_demand"][c,t] for competencies_2 in allowed_competency_combinations for c in competencies_2 for t in t_covered_by_shift[shift[0], shift[1]] if (c, t) in state.soft_vars["deviation_from_ideal_demand"] if state.soft_vars["deviation_from_ideal_demand"][c,t] < 0)
+        if(deviation_from_demand < 6):
+            print("WE: Deviation from demand: " + str(deviation_from_demand))
```
Comment (ekallevik) [inline heuristic/repair_operators.py:277]: Use `logger.info`

File: heuristic/repair_operators.py [hunk 9]
```diff
@@ -382,47 +360,47 @@ def worst_employee_regret_repair(competencies, t_covered_by_shift, employee_with
...
-
-        below_minimum_demand(state, destroy_set, employee_with_competencies, demand, competencies, t_covered_by_shift)
-        above_maximum_demand(state, destroy_set, employee_with_competencies, demand, competencies, t_covered_by_shift)
-        cover_multiple_demand_periods(state, destroy_set, t_covered_by_shift, competencies)
-        mapping_shift_to_demand(state, destroy_set, t_covered_by_shift, shifts_overlapping_t, competencies)
+        
+        # below_minimum_demand(state, destroy_set, employee_with_competencies, demand, competencies, t_covered_by_shift)
+    
+        # cover_multiple_demand_periods(state, destroy_set, t_covered_by_shift, competencies)
+        # mapping_shift_to_demand(state, destroy_set, t_covered_by_shift, shifts_overlapping_t, competencies)
```
Comment (ekallevik) [inline heuristic/repair_operators.py:367]: Remove or reintroduce lines 363-367

File: heuristic/state.py [hunk 10]
```diff
@@ -12,15 +12,32 @@ def __init__(self, decision_vars, soft_vars, hard_vars, objective_function_value
...
 
+        # self.cover_min_demand
+        # self.cover_max_demand
+        # self.one_demand_per_time
+        # self.one_shift_per_day
+        # self.one_weekly_off
+        # self.work_during_off
+        # self.shift_demand_map
+        # self.break_contracted_hours
+
```
Comment (ekallevik) [inline heuristic/state.py:33]: Remove or reintroduce line 15-21 and 25-32

File: preprocessing/shift_generation.py [hunk 11]
```diff
@@ -208,6 +208,8 @@ def get_shifts_per_day(shifts, days):
             if shift[0] >= 24 * int(day) + TIME_DEFINING_SHIFT_DAY:
                 if day == days[-1]:
                     shifts_per_day[day].append(shift)
+                    #if(shift not in shifts_in_week):
+                        #shifts_in_week[int(day/7)].append(shift)
```
Comment (ekallevik) [inline preprocessing/shift_generation.py:212]: Remove or reintroduce

File: heuristic/repair_operators.py (hunk 6 above)
Comment (hkgrov) [reply heuristic/repair_operators.py:75]: Done

**Pull request merged**